### DISCUSSION TITLE
OpenSSL 3.1.2 decoder provider fallback error

### DISCUSSION THREAD

**Discussion by Seulhui** [openssl/openssl#21882] [Q&A] [answered]
Hi everyone, 

I've implemented my own provider only including keymgnt and signature operations on OpenSSL 3.0.7, and it worked well. However, I found that it is much slower than the engine, even if it operates via the same APIs for my crypto module. I found that OpenSSL 3.0.X has a performance problem, so I updated the OpenSSL version to v3.1.2 and tried to test its performance. However, it showed a segmentation fault with the same provider implementation. It has no problem if it works only with sign operation; however, when I call X509_REQ_sign and  OCSP_request_sign, it shows segment faults.

#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7ad8548 in __GI_abort () at abort.c:79
#2  0x00007ffff7b2ea29 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7c34f02 "%s\n") at ../sysdeps/posix/libc_fatal.c:181
#3  0x00007ffff7b35b5a in malloc_printerr (str=str@entry=0x7ffff7c330a0 "free(): invalid pointer") at malloc.c:5332
#4  0x00007ffff7b374bc in _int_free (av=<optimized out>, p=<optimized out>, have_lock=0) at malloc.c:4173
#5  0x00007ffff76a9a3d in ossl_asn1_item_embed_free (pval=pval@entry=0x7fffffffe268, it=it@entry=0x7ffff7a53660 <local_it>, embed=embed@entry=0)
    at openssl3/crypto/asn1/tasn_fre.c:115
#6  0x00007ffff76a9c37 in ASN1_item_ex_free (pval=pval@entry=0x7fffffffe268, it=it@entry=0x7ffff7a53660 <local_it>) at openssl3/crypto/asn1/tasn_fre.c:25
#7  0x00007ffff76a8753 in asn1_item_ex_d2i_intern (propq=0x0, libctx=0x0, ctx=0x7fffffffe1d0, opt=0 '\000', aclass=0, tag=-1, it=0x7ffff7a53660 <local_it>, len=-1, in=0x7fffffffe288, pval=0x7fffffffe268)
    at openssl3/crypto/asn1/tasn_dec.c:121
#8  ASN1_item_d2i_ex (pval=pval@entry=0x7fffffffe268, in=in@entry=0x7fffffffe288, len=len@entry=-1, it=0x7ffff7a53660 <local_it>, libctx=libctx@entry=0x0, propq=propq@entry=0x0)
    at openssl3/crypto/asn1/tasn_dec.c:144
#9  0x00007ffff76a876b in ASN1_item_d2i (pval=pval@entry=0x7fffffffe268, in=in@entry=0x7fffffffe288, len=len@entry=-1, it=<optimized out>)
    at openssl3/crypto/asn1/tasn_dec.c:154
#10 0x00007ffff76ac539 in d2i_X509_ALGOR (a=a@entry=0x7fffffffe268, in=in@entry=0x7fffffffe288, len=len@entry=-1) at openssl3/crypto/asn1/x_algor.c:27
#11 0x00007ffff76979e3 in ASN1_item_sign_ctx (it=it@entry=0x7ffff7a77d40 <local_it>, algor1=<optimized out>, algor1@entry=0x555555640d68, algor2=<optimized out>, algor2@entry=0x0,
    signature=signature@entry=0x555555640e00, data=data@entry=0x555555640d30, ctx=ctx@entry=0x555555688d60) at openssl3/crypto/asn1/a_sign.c:198
#12 0x00007ffff7697d31 in ASN1_item_sign_ex (it=0x7ffff7a77d40 <local_it>, algor1=algor1@entry=0x555555640d68, algor2=algor2@entry=0x0, signature=signature@entry=0x555555640e00,
    data=data@entry=0x555555640d30, id=id@entry=0x0, pkey=0x555555644110, md=0x7ffff7a661c0 <sha256_md>, libctx=0x0, propq=0x0) at openssl3/crypto/asn1/a_sign.c:143
#13 0x00007ffff791e852 in X509_REQ_sign (x=0x555555640d30, pkey=0x555555644110, md=0x7ffff7a661c0 <sha256_md>) at openssl3/crypto/x509/x_all.c:119

asn1_item_embed_d2i returns not bigger than 0 value, it tries to free via ASN1_item_ex_free, and it makes segfault. In my implementation, I loaded not only my provider for keymgmt and signature but also the "default" provider to use fallback operations, including decoder. I guess that it doesn't work properly, not like OpenSSL 3.0.1.  Is there anyone the reason for this change between the two OpenSSL versions?

Regards,
Seulhui Lee.

**Comment by t8m** [comment]
Unfortunately there is too little information as for what are you doing in your provider. In the backtrace above there is no provider code in play.

**Reply by Seulhui** [reply]
My provider operates the loading key from my key storage module and signs data via the loaded private key and my crypto module. The EVP_PKEY, which is one of the parameters of X509_REQ_sign, was loaded from my keymgmt provider, and when I called X509_REQ_sign, my provider's signature init function was called, but it failed during the signing operation like the backtrace.

**Reply by mattcaswell** [reply]
> my provider's signature init function was called,

Hmm...but that call does not appear to be in the above stack trace. Did your init function succeed or fail? By "init" I am assuming you mean the init for the digest sign operation?

From the stack trace it seems to be failing in `d2i_X509_ALGOR` which is passed the Algorithm ID as an octet string which it has obtained from the provider via a call to `EVP_PKEY_CTX_get_params()`. It would have tried to obtain the param `OSSL_SIGNATURE_PARAM_ALGORITHM_ID`. Do you see your provider responding to the request for that param? And does it supply a valid response?

The `d2i_X509_ALGOR` function is trying to reuse the X509_ALGOR data in the `sig_alg` field of the `X509_REQ` that you pass to the `X509_REQ_sign()` function. How did you construct the `X509_REQ`?

**Reply by Seulhui** [reply]
Thank you for the kind answers. I found that the algorithm ID residents in the stack(I don't know why it works for OpenSSL3.0.7), so I resolved the problem after allocating it to the heap.

**Comment by nhorman** [comment, marked as answer]
These 3 lines:
```
#3 0x00007ffff7b35b5a in malloc_printerr (str=str@entry=0x7ffff7c330a0 "free(): invalid pointer") at malloc.c:5332
#4 0x00007ffff7b374bc in _int_free (av=, p=, have_lock=0) at malloc.c:4173
#5 0x00007ffff76a9a3d in ossl_asn1_item_embed_free (pval=pval@entry=0x7fffffffe268, it=it@entry=0x7ffff7a53660 <local_it>, embed=embed@entry=0)
```
I think are telling. ossl_asn1_item_embed_free on line 115 frees *pval, which looking up the stack is passed in as a pointer to a pointer to an <X509_ALGOR type>->sig_alg

While pval itself looks valid (its a stack address), whatever its pointing to is garbage (or NULL).  My guess would be that your provider is providing the X509_ALGOR implementation and you didn't setup sig_alg properly.

**Reply by Seulhui** [reply]
This was the answer!!! Thank you for your insight. :) I fixed the problem.

**Comment by slontis** [comment]
If you care about speed, and you are passing EVP_sha256() to multiple function calls, you may want to fetch instead.
See doc/man7/ossl-guide-libcrypto-introduction.pod

**Reply by Seulhui** [reply]
I'll see the fetch APIs. Thank you for sharing it. :)
